source/extensions/filters/network/thrift_proxy/binary_protocol.cc

```diff
@@ -0,0 +1,300 @@
+                                      FieldType& value_type, uint32_t& size) {
+  // Minimum length:
+  //   key type: 1 byte +
+  //   value type: 1 byte +
+  //   map size: 4 bytes
+  if (buffer.length() < 6) {
+    return false;
+  }
+
+  FieldType ktype = static_cast<FieldType>(BufferHelper::peekI8(buffer, 0));
+  FieldType vtype = static_cast<FieldType>(BufferHelper::peekI8(buffer, 1));
+  int32_t s = BufferHelper::peekI32(buffer, 2);
+  if (s < 0) {
+    throw EnvoyException(fmt::format("negative binary protocol map size {}", s));
+  }
+
+  buffer.drain(6);
+
+  key_type = ktype;
+  value_type = vtype;
+  size = static_cast<uint32_t>(s);
+
+  return true;
+}
+
+bool BinaryProtocolImpl::readMapEnd(Buffer::Instance& buffer) {
+  UNREFERENCED_PARAMETER(buffer);
+  return true;
+}
+
+bool BinaryProtocolImpl::readListBegin(Buffer::Instance& buffer, FieldType& elem_type,
+                                       uint32_t& size) {
+  // Minimum length:
+  //   elem type: 1 byte +
+  //   map size: 4 bytes
+  if (buffer.length() < 5) {
+    return false;
+  }
+
+  FieldType type = static_cast<FieldType>(BufferHelper::peekI8(buffer));
+  int32_t s = BufferHelper::peekI32(buffer, 1);
+  if (s < 0) {
+    throw EnvoyException(fmt::format("negative binary protocol list/set size {}", s));
+  }
+  buffer.drain(5);
+
+  elem_type = type;
+  size = static_cast<uint32_t>(s);
+
+  return true;
+}
+
+bool BinaryProtocolImpl::readListEnd(Buffer::Instance& buffer) {
+  UNREFERENCED_PARAMETER(buffer);
+  return true;
+}
+
+bool BinaryProtocolImpl::readSetBegin(Buffer::Instance& buffer, FieldType& elem_type,
+                                      uint32_t& size) {
+  return readListBegin(buffer, elem_type, size);
+}
+
+bool BinaryProtocolImpl::readSetEnd(Buffer::Instance& buffer) { return readListEnd(buffer); }
+
+bool BinaryProtocolImpl::readBool(Buffer::Instance& buffer, bool& value) {
+  if (buffer.length() < 1) {
+    return false;
+  }
+
+  value = BufferHelper::drainI8(buffer) != 0;
+  return true;
+}
+
+bool BinaryProtocolImpl::readByte(Buffer::Instance& buffer, uint8_t& value) {
+  if (buffer.length() < 1) {
+    return false;
+  }
+  value = BufferHelper::drainI8(buffer);
+  return true;
+}
+
+bool BinaryProtocolImpl::readInt16(Buffer::Instance& buffer, int16_t& value) {
+  if (buffer.length() < 2) {
+    return false;
+  }
+  value = BufferHelper::drainI16(buffer);
+  return true;
+}
+
+bool BinaryProtocolImpl::readInt32(Buffer::Instance& buffer, int32_t& value) {
+  if (buffer.length() < 4) {
+    return false;
+  }
+  value = BufferHelper::drainI32(buffer);
+  return true;
+}
+
+bool BinaryProtocolImpl::readInt64(Buffer::Instance& buffer, int64_t& value) {
+  if (buffer.length() < 8) {
+    return false;
+  }
+  value = BufferHelper::drainI64(buffer);
+  return true;
+}
+
+bool BinaryProtocolImpl::readDouble(Buffer::Instance& buffer, double& value) {
+  ASSERT(sizeof(double) == sizeof(uint64_t));
```

Comment:
Since Envoy is using C++14, I propose using the new `static_assert` here, since that will work for both debug and non-debug builds. (https://en.cppreference.com/w/cpp/language/static_assert)

Reply:
✅